core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala

```diff
@@ -381,12 +388,85 @@ private[spark] class TaskSchedulerImpl(
 
   /**
    * Check whether the resources from the WorkerOffer are enough to run at least one task.
+   * Returns None if the resources don't meet the task requirements, otherwise returns
+   * the task resource assignments to give to the next task. Note that the assignments maybe
+   * be empty if no custom resources are used.
    */
-  private def resourcesMeetTaskRequirements(resources: Map[String, Buffer[String]]): Boolean = {
-    val resourcesFree = resources.map(r => r._1 -> r._2.length)
-    val meetsReqs = ResourceUtils.resourcesMeetRequirements(resourcesFree, resourcesReqsPerTask)
-    logDebug(s"Resources meet task requirements is: $meetsReqs")
-    meetsReqs
+  private def resourcesMeetTaskRequirements(
+      taskSet: TaskSetManager,
+      availCpus: Int,
+      availWorkerResources: Map[String, Buffer[String]]
+      ): Option[Map[String, ResourceInformation]] = {
+    val rpId = taskSet.taskSet.resourceProfileId
+    val taskCpus = sc.resourceProfileManager.taskCpusForProfileId(rpId)
+    // check if the ResourceProfile has cpus first since that is common case
+    if (availCpus < taskCpus) return None
+
+    val taskSetProf = sc.resourceProfileManager.resourceProfileFromId(rpId)
+    // remove task cpus since we checked already
+    val tsResources = taskSetProf.taskResources.filterKeys(!_.equals(ResourceProfile.CPUS))
+    val localTaskReqAssign = HashMap[String, ResourceInformation]()
+    if (tsResources.isEmpty) return Some(localTaskReqAssign.toMap)
+    // we go through all resources here so that we can make sure they match and also get what the
+    // assignments are for the next task
+    for ((rName, taskReqs) <- tsResources) {
+      val taskAmount = taskSetProf.getSchedulerTaskResourceAmount(rName)
+      availWorkerResources.get(rName) match {
+        case Some(workerRes) =>
+          val workerAvail = availWorkerResources.get(rName).map(_.size).getOrElse(0)
+          if (workerAvail >= taskAmount) {
+            localTaskReqAssign.put(rName, new ResourceInformation(rName,
+              workerRes.take(taskAmount).toArray))
+          } else {
+            return None
+          }
+        case None => return None
+      }
+    }
+    Some(localTaskReqAssign.toMap)
+  }
+
+  // Use the resource that the resourceProfile has as the limiting resource to calculate the
+  // total number of slots available based on the current offers.
+  private def calculateAvailableSlots(
```

Comment:
Maybe we can pass in the target slots and stop calculation once we reach the target.

Reply:
we could. the code would become a  loop (adding up as you went) and not like a map.sum().  honestly I just did what was there before and was thinking that barrier doesn't support dynamic allocation at this point (unless it got checked in without me seeing it?) so the number of executors is more likely to be less then what you need, not more.  
Do you see other cases?
I guess  when barrier scheduling starts to support dynamic allocation this could be more of an issue if you run a large ETL stage and then to smaller number with ML. 
I'm fine either way, if you think its an issue now I will change or we could wait til dynamic allocation is supported, what are you thoughts?  (note that I am changing that logic as it had a bug that Mridul found in his review)

Reply:
> unless it got checked in without me seeing it?

Barrier mode doesn't support dynamic allocation yet.


But sorry I don't quite understand your concern here.

With my understanding, `calculateAvailableSlots` only gives a chance to barrier stage to try to launch all tasks at the same time. But it never guarantee that barrier stage can be successfully launched even if there's enough slots(e.g. due to delay scheduling). So, I mean, whether we calculate total or partial sum of resource slots, it should makes no difference for barrier stage scheduling.

Reply:
Barrier just needs to make sure there are X slots available, if there are not it skips that round of scheduling. 
What I'm saying is this is how the code did it before (calculate all slots) and in the scenario people are using barrier scheduling, they can't use dynamic allocation, so I would  expect the user to request X executors up front so it is very unlikely the job would have X + Y executors, so even though calculateAvailableSlots would add up all slots, I wouldn't expect that to be more then X. Meaning calculateAvailableSlots isn't wasting any time adding in extra slots because they wouldn't be there.

Reply:
Ok, I see.